### ISSUE TITLE
gitlearnbrancing places too much emphasis on a solution's 'commit number'

### ISSUE THREAD

**Issue by Poddster** [pcottle/learnGitBranching#448]
The "commit number" appears to be specific when solving a solution. For example, in intro4, learngitbranching claims '6' is the best solution:

````bash
$ git checkout -b bugFix
$ git commit
$ git checkout master
$ git commit
$ git checkout bugFix
$ git rebase master
````

But you can do the same in 5:

````bash
$ git branch bugFix
$ git commit
$ git checkout bugFix
$ git commit
$ git rebase master
````

So from a "puzzle game" POV this is a bit annoying. But I also believe, from a pedagogical POV, it's teaching people things that aren't entirely true about git! 

A "commit" in git is actually a {state of the tree, parent commits, commit message, timestamp, etc} object, and an sha1 id labels this commit object. 

So in terms of "real git" for the 5-step solution to intro4 the state of the working tree at this point will be identical, as will the commit graph, the only difference is that the commits will have different timestamps and sha1 ids. **But the timestamps and sha1 ids are unimportant!** That's exactly what happens in rebases and cherry-picks etc -- the commit-tree stays the same, as do the commit messages, but parent pointers and timestamps change, which means a new commit sha1.

I'd say it's an important concept in git is that the sha1 id is relatively unimportant and that you're free to change your graph however you like until you push. And I'd like learngitbranching to teach that to people. But learngitbrancing's rigid enforcement of having the exact same commit numbers means it's not teaching "real" practices and real git problem solving methods.

tl;dr:

I'd like to see the solutions be considered 'acceptable' based on the 'shape' of the graph, rather than on having the exact commit number (aka sha1).

**Comment by pcottle** [comment]
Well to be super technical, you can do it in 4 commands:
```
$ git commit
$ git checkout -b bugFix C1
$ git commit
$ git rebase master
```

But that's besides the point. For this particular level yes, I examine all branches (and HEAD) and the exact refs they point to. but thats not the only comparison mode I have:
https://github.com/pcottle/learnGitBranching/blob/master/src/js/graph/treeCompare.js#L49

Many levels do hashAgnostic comparison, where they don't compare which version of a commit you have, but simply that you have it at all.

I totally agree that theres a difference between reality (where sha1's are purely markers/identifiers) and this educational tool. You can achieve the same logical steps in sometimes a shorter fashion, but with different sha1 ID's afterwards.

The issue is that I have to distinguish these "smarter and more efficient" solutions from invalid and "I don't know what I'm doing" states. I thought about only allowing the exact sequence of the solution but that seemed too restrictive. I also thought about tracking if the given commands had been run *at all*, but its also quite easy to beat that.

Going with the sha1 IDs + shape of the graph was the best compromise I could make. I have to use something more specific than *just* the shape of the graph, since for some levels, you already start in the desired shape of the graph (ignoring commits no longer accessible via a branch):
![screenshot 2017-12-21 at 10 04 49 am](https://user-images.githubusercontent.com/1135007/34268661-7292da6a-e636-11e7-96f6-c5efbb553ca9.png)

or you even start in the desired shape in the desired order of the graph:
![screenshot 2017-12-21 at 10 10 58 am](https://user-images.githubusercontent.com/1135007/34268891-40652ea2-e637-11e7-9afd-3d1a27333d87.png)

so I have to track some kind of edit-or-amend history of commits, and the IDs + a dash for each edit is where I ended up.

I'd also like to note that most people using this tutorial are really intro-level beginners, I don't think their main takeaway will be that the sha1 IDs are important but rather the high level concepts and which commands to use when. If this app is insufficient for your educational purposes then I apologize, but I had to go with some kind of compromise of how to "grade" the given solution.
